en/lessons/basics/pattern-matching.md

```diff
@@ -1,107 +1,426 @@
+functionName RecordDataConstructor2{ field1 = localName1 } = (expression2)
+functionName RecordDataConstructor3
+    { field1 = localName1, field2 = localName2, ... }      = (expression3)
+```
+
+### Example Record Type
+
+Say we have a this record type
+
+```haskell
+data Vehicle
+    = Aircraft
+        { vaColor :: Text
+        , vaWings :: Int
+        }
+    | Boat
+        { vbColor :: Text 
+        , vbSails :: Int
+        }
+    | Car 
+        { vcColor  :: Text
+        , vcWheels :: Int
+        }
+    deriving (Eq, Show)
+```
+
+### Pattern Matching the Record Constructor
+
+```haskell
+-- Can this vehicle fly in the air?
+-- Even though we only want to match on the data constructor, we have to write
+-- an `_` for each field in the record
+vehicleCanFlyExplicit :: Vehicle -> Bool
+vehicleCanFlyExplicit (Aircraft _ _) = True
+vehicleCanFlyExplicit _              = False
+```
+
+A record type could potentially have a hundred fields, and we would not want to
+write an `_` for each one, so there is a better way
+
+```haskell
+-- Same function as above, with a simpler syntax
+-- Note how the curly braces indicate matching on the data constructor, but 
+-- nothing after
+vehicleCanFly :: Vehicle -> Bool
+vehicleCanFly Aircraft{} = True
+vehicleCanFly _          = False
+```
+
+The empty curly braces (`{}`) mean to fill in the rest of the pattern with
+catch-all underscores (`_`)
+
+### Pattern Matching with Record Fields
+
+Curly braces can also be used to read fields of the record
+
+```haskell
+-- What color is this vehicle?
+-- Note how the contents of a field are given a name, which is then used 
+-- in the function definition
+vehicleColor :: Vehicle -> Text
+vehicleColor Aircraft { vaColor = color } = "A " <> color <> " aircraft"
+vehicleColor Boat     { vbColor = color } = "A " <> color <> " boat"
+vehicleColor Car      { vcColor = color } = "A " <> color <> " car"
+```
+
+It is also possible to use more than one field
+
+```haskell
+-- What kind of flying vehicle is this?
+-- Note how more than one record field is used in the function definition
+flyingVehicleType :: Vehicle -> Text
+flyingVehicleType Aircraft 
+    { vaColor = color, vaWings = wingQty }
+    | wingQty == 0 = colorText <> "Zeppelin"
+    | wingQty == 1 = colorText <> "Stealth Plane"
+    | wingQty == 2 = colorText <> "Airplane"
+    | wingQty  > 2 = colorText <> "Helicopter"
+    | otherwise    = colorText <> "UFO"
+  where
+    colorText = "A " <> color <> " "
+flyingVehicleType _ = "This vehicle can't fly" 
+```
+
+### Record Pattern Matching Conveniences
+
+Two language extensions can add convenience to working with records and patterns
+
+1. `RecordWildCards` can automaticlly define local names that exactly match the
+   record field names
+2. `NamedFieldPuns` allows you to just input the record field name, and have it
+   defined as a local name
+
+Feel free the explore these further once you feel confident writing record
+patterns by hand
```

Comment:
A period is missing. :)

Reply:
None of the one-line statements have periods. I think it looks a bit nicer this way, should I go and add periods to them?

Reply:
I think that from a typographical point of view it's better, yes. :)